state/cleanup.go

```diff
@@ -1433,6 +1436,63 @@ func (st *State) cleanupForceRemoveMachine(machineId string, cleanupArgs []bson.
 	return machine.Remove()
 }
 
+// cleanupEvacuateMachine is initiated by machine.Destroy() to gracefully remove units
+// from the machine before then kicking off machine destroy.
+// Currently only evacuates juju system charms like the controller charm.
+func (st *State) cleanupEvacuateMachine(machineId string, cleanupArgs []bson.Raw) error {
+	if len(cleanupArgs) > 0 {
+		return errors.Errorf("expected no arguments, got %d", len(cleanupArgs))
+	}
+
+	machine, err := st.Machine(machineId)
+	if errors.IsNotFound(err) {
+		return nil
+	} else if err != nil {
+		return errors.Trace(err)
+	}
+	if machine.Life() != Alive {
+		return nil
+	}
+
+	units, err := machine.Units()
```

Comment:
We'd need to check that we are just removing the "juju-" units just to be save. We could get here and someone has assigned another unit to the machine. We could pass in the units to remove as args, or just filter.

Reply:
This check is done in the Destroy() call that creates the evacuate job and now prevents addition of extra units.